File: app/src/main/java/com/bryan/platform/network/RetrofitClient.java
package com.bryan.platform.network;

import androidx.annotation.NonNull;

import com.bryan.platform.utils.SessionManager;

import java.util.Arrays;
import java.util.List;
import java.util.concurrent.TimeUnit;

import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

/**
 * Retrofit 客户端单例，用于创建 API 服务实例
 * 已整合认证拦截器和会话管理
 */
public final class RetrofitClient {

    // 后端 API 的基础 URL
    private static final String BASE_URL = "http://10.0.2.2:8080/";

    // 不需要认证的公共 API 路径列表
    private static final List<String> PUBLIC_URLS = Arrays.asList(
    "/api/auth/login",
    "/api/auth/register"
    );

    private volatile static RetrofitClient INSTANCE;
    private final Retrofit retrofit;

    private RetrofitClient() {
        // 日志拦截器
        HttpLoggingInterceptor loggingInterceptor = new HttpLoggingInterceptor();
        loggingInterceptor.setLevel(HttpLoggingInterceptor.Level.BODY);

        // OkHttpClient
        OkHttpClient okHttpClient = new OkHttpClient.Builder()
            .addInterceptor(new AuthInterceptor())
            .addInterceptor(loggingInterceptor)
            .connectTimeout(30, TimeUnit.SECONDS)
            .readTimeout(30, TimeUnit.SECONDS)
            .writeTimeout(30, TimeUnit.SECONDS)
            .build();

        // Retrofit
        retrofit = new Retrofit.Builder()
            .baseUrl(BASE_URL)
            .client(okHttpClient)
            .addConverterFactory(GsonConverterFactory.create())
            .build();
    }

    public static RetrofitClient getInstance() {
        if (INSTANCE == null) {
            synchronized (RetrofitClient.class) {
                if (INSTANCE == null) {
                    INSTANCE = new RetrofitClient();
                }
            }
        }
        return INSTANCE;
    }

    /**
     * 创建指定 API 服务的实例
     */
    public <T> T createService(@NonNull Class<T> serviceClass) {
        return retrofit.create(serviceClass);
    }

    /* ========================= 内部类：认证拦截器 ========================= */

    private static class AuthInterceptor implements Interceptor {

        @NonNull
        @Override
        public Response intercept(@NonNull Chain chain) throws java.io.IOException {
            Request original = chain.request();
            String path = original.url().encodedPath();

            // 公共接口直接放行
            boolean isPublic = false;
            for (String publicPath : PUBLIC_URLS) {
            if (path.endsWith(publicPath)) {
                isPublic = true;
                break;
            }
        }
            if (isPublic) {
                return chain.proceed(original);
            }

            // 非公共接口：需要 token
            String token = SessionManager.getInstance().fetchAuthToken();
            if (token != null) {
                Request newRequest = original.newBuilder()
                    .header("Authorization", "Bearer " + token)
                    .build();
                return chain.proceed(newRequest);
            }

            // 没有 token 也放行（由后端决定 401/403）
            return chain.proceed(original);
        }
    }
}

File: app/src/main/java/com/bryan/platform/network/RetrofitClient.kt
package com.bryan.platform.network

import com.bryan.platform.utils.SessionManager
import okhttp3.Interceptor
import okhttp3.OkHttpClient
import okhttp3.Response
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.util.concurrent.TimeUnit

/**
 * Retrofit 客户端单例对象，用于创建 API 服务实例
 * 已整合认证拦截器和会话管理
 */
object RetrofitClient {

    // 后端 API 的基础 URL
    private const val BASE_URL = "http://10.0.2.2:8080/" // 模拟器访问宿主机器的本地服务地址

    // 不需要认证的公共 API 路径列表
    private val PUBLIC_URLS = listOf(
        "/api/auth/login",
        "/api/auth/register"
    )

    // 认证拦截器，用于自动添加 Token 到请求头
    private class AuthInterceptor : Interceptor {
        override fun intercept(chain: Interceptor.Chain): Response {
            val originalRequest = chain.request()
            val requestPath = originalRequest.url.encodedPath // 获取请求路径

            // 检查当前请求路径是否在公共 URL 列表中
            val isPublicUrl = PUBLIC_URLS.any { publicPath ->
                requestPath.endsWith(publicPath) // 检查请求路径是否以公共路径结尾
            }

            // 如果是公共 URL，则不添加 Token，直接继续请求
            if (isPublicUrl) {
                return chain.proceed(originalRequest)
            }

            // 如果不是公共 URL，则尝试添加 Token
            val token = SessionManager.getInstance().fetchAuthToken()

            return if (token != null) {
                // 如果 Token 存在，添加到请求头
                val newRequest = originalRequest.newBuilder()
                    .header("Authorization", "Bearer $token")
                    .build()
                chain.proceed(newRequest)
            } else {
                // 没有 Token 则直接继续请求 (对于需要认证但没有 Token 的情况)
                chain.proceed(originalRequest)
            }
        }
    }

    // OkHttpClient 配置
    private val okHttpClient: OkHttpClient by lazy {
        // 创建 HttpLoggingInterceptor 实例，用于记录网络请求和响应的日志
        val loggingInterceptor = HttpLoggingInterceptor().apply {
            // 设置日志级别为 BODY，表示记录请求和响应头以及正文
            level = HttpLoggingInterceptor.Level.BODY
        }

        OkHttpClient.Builder()
            .addInterceptor(AuthInterceptor()) // 添加认证拦截器
            .addInterceptor(loggingInterceptor) // 添加日志拦截器
            .connectTimeout(30, TimeUnit.SECONDS) // 设置连接超时
            .readTimeout(30, TimeUnit.SECONDS)    // 设置读取超时
            .writeTimeout(30, TimeUnit.SECONDS)   // 设置写入超时
            .build()
    }

    // Retrofit 实例，懒加载
    private val retrofit: Retrofit by lazy {
        Retrofit.Builder()
            .baseUrl(BASE_URL) // 设置基础 URL
            .client(okHttpClient) // 设置自定义的 OkHttpClient
            .addConverterFactory(GsonConverterFactory.create()) // 添加 Gson 转换器
            .build()
    }

    /**
     * 创建指定 API 服务的实例
     * @param serviceClass 接口类的 Class 对象
     * @return API 服务的实例
     */
    fun <T> createService(serviceClass: Class<T>): T {
        return retrofit.create(serviceClass)
    }

    /**
     * Kotlin 扩展函数方式创建服务 (更简洁)
     */
    inline fun <reified T> createService(): T {
        return createService(T::class.java)
    }
}
